kj-rs/awaiter.rs

```diff
@@ -46,52 +46,43 @@ impl<Data: std::marker::Unpin> PromiseAwaiter<Data> {
-
-        // Safety: `awaiter` stores `rust_waker_ptr` and uses it to call `wake()`. Note that
-        // `awaiter` is `self.awaiter`, which lives before `self.option_waker`. Since struct members
-        // are dropped in declaration order, the `rust_waker_ptr` that `awaiter` stores will always
-        // be valid during its lifetime.
-        //
-        // We pass a mutable pointer to C++. This is safe, because our use of the OptionWaker inside
-        // of `std::task::Waker` is synchronized by ensuring we only allow calls to `poll()` on the
-        // thread with the Promise's event loop active.
-        let rust_waker_ptr = &raw mut self.as_mut().option_waker;
-
-        // Safety:
-        // 1. We do not implement Unpin for PromiseAwaiter.
-        // 2. Our Drop trait implementation does not move the awaiter value, nor do we use
-        //    `repr(packed)` anywhere.
-        // 3. The backing memory is inside our pinned Future, so we can be assured our Drop trait
-        //    implementation will run before Rust re-uses the memory.
-        //
-        // https://doc.rust-lang.org/std/pin/index.html#choosing-pinning-to-be-structural-for-field
-        let awaiter = unsafe { self.map_unchecked_mut(|s| &mut s.awaiter) };
-
-        // Safety:
-        // 1. We trust that LazyPinInit's implementation passed us a valid pointer to an
-        //    uninitialized GuardedRustPromiseAwaiter.
-        // 2. We do not read or write to the GuardedRustPromiseAwaiter's memory, so there are no atomicity
-        //    nor interleaved pointer reference access concerns.
-        //
-        // https://doc.rust-lang.org/std/ptr/index.html#safety
-        let result = awaiter.get_or_init(move |ptr: *mut GuardedRustPromiseAwaiterRepr| unsafe {
-            crate::ffi::guarded_rust_promise_awaiter_new_in_place(
-                ptr as *mut GuardedRustPromiseAwaiter,
-                rust_waker_ptr,
-                node.expect("node should be Some in call to init()"),
-            );
-        });
-
-        let awaiter = unsafe {
-            let raw = Pin::into_inner_unchecked(result) as *mut GuardedRustPromiseAwaiterRepr;
+        // Safety: We never move out of `this`.
+        let this = unsafe { Pin::into_inner_unchecked(self.as_mut()) };
+
+        // Initialize the awaiter if not already done
+        if !this.awaiter_initialized {
+            // On our first invocation, `node` will be Some, and `get_awaiter` will forward its
+            // contents into GuardedRustPromiseAwaiter's constructor. On all subsequent invocations, `node`
+            // will be None and the constructor will not run.
+            let node = this.node.take();
+
+            // Safety: `awaiter` stores `rust_waker_ptr` and uses it to call `wake()`. Note that
+            // `awaiter` is `this.awaiter`, which lives before `this.option_waker`. Since struct members
+            // are dropped in declaration order, the `rust_waker_ptr` that `awaiter` stores will always
```

Comment:
Nit: This comment, as well as the comment above the `option_waker` field in PromiseAwaiter's struct definition on lines 28-29, are now a bit out-of-date: MaybeUninit needs manually dropping, so our manually-implemented Drop trait is what guarantees that the value in `this.awaiter` is destroyed before `this.option_waker` now.